bootstore/src/schemes/v0/peer_networking.rs

```diff
@@ -0,0 +1,635 @@
+    std::io::Write::write(&mut cursor, &[0u8; FRAME_HEADER_SIZE]);
+    cursor.set_position(FRAME_HEADER_SIZE as u64);
+    ciborium::into_writer(msg, &mut cursor)?;
+    let size: u32 =
+        (cursor.position() - FRAME_HEADER_SIZE as u64).try_into().unwrap();
+    let mut buf = cursor.into_inner();
+    buf[0..FRAME_HEADER_SIZE].copy_from_slice(&size.to_be_bytes());
+    Ok(buf)
+}
+
+// Decode the 4-byte big-endian frame size header
+fn read_frame_size(buf: [u8; FRAME_HEADER_SIZE]) -> usize {
+    u32::from_be_bytes(buf) as usize
+}
+
+#[derive(Debug, From)]
+enum HandshakeError {
+    Serialization(ciborium::ser::Error<std::io::Error>),
+    Deserialization(ciborium::de::Error<std::io::Error>),
+    Io(tokio::io::Error),
+    UnsupportedScheme,
+    UnsupportedVersion,
+    Timeout,
+}
+
+// Perform scheme/version negotiation and exchange peer_ids for scheme v0
+async fn perform_handshake(
+    sock: TcpStream,
+    local_peer_id: &Baseboard,
+    local_addr: SocketAddrV6,
+    log: &Logger,
+) -> Result<(OwnedReadHalf, OwnedWriteHalf, Identify), HandshakeError> {
+    // Enough to hold the `Hello` and `Identify` messages
+    let mut read_buf = [0u8; 128];
+    let (mut read_sock, mut write_sock) = sock.into_split();
+
+    // Serialize and write the handshake messages into `write_buf`
+    let mut hello_cursor = Cursor::new(Hello::default().serialize());
+    let identify = write_framed(&Identify {
+        id: local_peer_id.clone(),
+        addr: local_addr,
+    })?;
+    let mut identify_cursor = Cursor::new(&identify);
+
+    let handshake_start = Instant::now();
+
+    // Read `Hello` and the frame size of `Identify`
+    let initial_read = Hello::serialized_size() + FRAME_HEADER_SIZE;
+
+    let mut total_read = 0;
+    let mut identify_len = 0;
+    let mut identify: Option<Identify> = None;
+
+    loop {
+        let timeout =
+            KEEPALIVE_TIMEOUT.saturating_sub(Instant::now() - handshake_start);
+
+        let end = initial_read + identify_len;
+
+        let hello_written = !hello_cursor.has_remaining();
+
+        if identify.is_some() && !identify_cursor.has_remaining() {
+            return Ok((read_sock, write_sock, identify.unwrap()));
+        }
+
+        tokio::select! {
+            _ = sleep(timeout) => {
+                return Err(HandshakeError::Timeout);
+            }
+            _ = write_sock.writable(), if identify_cursor.has_remaining() => {
+                if hello_cursor.has_remaining() {
+                    write_sock.write_buf(&mut hello_cursor).await?;
+                } else {
+                    write_sock.write_buf(&mut identify_cursor).await?;
+                }
+            }
+            res = read_sock.read(&mut read_buf[total_read..end]), if identify.is_none() => {
+                let n = res?;
+                total_read += n;
+                if total_read < initial_read {
+                    continue;
+                }
+                if total_read == initial_read {
```

Comment:
This section is a little too sneaky for my taste – it requires noticing that `read_sock.read(&mut read_buf[total_read..end])` will only read up to `end` (so we can do an exact comparison `total_read == initial_read`), **and** `end` changes once we've read the `initial_len` of the data and decoded `identify_len`.

What do you think about splitting this function into separate sections? I mocked it up locally (didn't test it, but it compiles fine), along with other recommendations:
```rust
    // Serialize and write the handshake + identity messages to a local buffer
    let out: Vec<u8> = Hello::default()
        .serialize()
        .into_iter()
        .chain(write_framed(&Identify {
            id: local_peer_id.clone(),
            addr: local_addr,
        })?)
        .collect();
    let mut out_cursor = Cursor::new(&out);

    let handshake_start = Instant::now();
    let get_timeout =
        || KEEPALIVE_TIMEOUT.saturating_sub(Instant::now() - handshake_start);

    // Read `Hello` and the frame size of `Identify`
    const INITIAL_READ: usize = Hello::serialized_size() + FRAME_HEADER_SIZE;

    // Read the initial Hello message and the length of the following Identify
    let mut read_pos = 0;
    let identify_len = loop {
        let timeout = get_timeout();
        tokio::select! {
            _ = sleep(timeout) => {
                return Err(HandshakeError::Timeout);
            }
            _ = write_sock.writable(), if out_cursor.has_remaining() => {
                write_sock.write_buf(&mut out_cursor).await?;
            }
            res = read_sock.read(&mut read_buf[read_pos..INITIAL_READ]) => {
                read_pos += res?;
                if read_pos == INITIAL_READ {
                    let hello =
                        Hello::from_bytes(&read_buf[..Hello::serialized_size()]).unwrap();
                    if hello.scheme != 0 {
                        return Err(HandshakeError::UnsupportedScheme);
                    }
                    if hello.version != 0 {
                        return Err(HandshakeError::UnsupportedVersion);
                    }
                    break read_frame_size(
                        read_buf[Hello::serialized_size()..INITIAL_READ]
                            .try_into()
                            .unwrap(),
                    );
                }
            }
        }
    };

    // Read and decode the Identify message, using the length found above
    let mut read_pos = 0;
    let identify: Identify = loop {
        let timeout = get_timeout();
        tokio::select! {
            _ = sleep(timeout) => {
                return Err(HandshakeError::Timeout);
            }
            _ = write_sock.writable(), if out_cursor.has_remaining() => {
                write_sock.write_buf(&mut out_cursor).await?;
            }
            res = read_sock.read(&mut read_buf[read_pos..identify_len]) => {
                read_pos += res?;
                if read_pos == identify_len {
                    break ciborium::from_reader(&read_buf[..identify_len])?;
                }
            }
        }
    };

    // Flush the remaining output from out_cursor
    while !out_cursor.has_remaining() {
        let timeout = get_timeout();
        tokio::select! {
            _ = sleep(timeout) => {
                return Err(HandshakeError::Timeout);
            }
            _ = write_sock.writable() => {
                write_sock.write_buf(&mut out_cursor).await?;
            }
        }
    }
    Ok((read_sock, write_sock, identify))
```

Reply:
This doesn't pass tests. The handshake doesn't complete. While I agree this structure is a touch clearer, I don't feel like debugging right now. I took all your other suggestions for the handshake though.

Reply:
ah, the issue is probably `while !out_cursor.has_remaining()` (should be `while out_cursor.has_remaining()`); up to you if you want to re-test.